pkg/operator/admission/webhook.go

```diff
@@ -0,0 +1,203 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"os"
+)
+
+const (
+	appName            = "rook-admission-controller"
+	secretVolumeName   = "webhook-certs"
+	secretMountPath    = "/etc/webhook"
+	serviceAccountName = "rook-admission-controller-sa"
+	storageFlag        = "ceph"
+)
+
+var (
+	validatePath       = "/validate"
+	servicePort  int32 = 443
+	serverPort   int32 = 8079
+	namespace          = os.Getenv(k8sutil.PodNamespaceEnvVar)
+)
+
+func isSecretPresent(context *clusterd.Context) (bool, error) {
+	_, err := context.Clientset.CoreV1().Secrets(namespace).Get(appName, metav1.GetOptions{})
+	if err != nil {
+		// If secret is not found. All good ! Proceed with rook without admission controllers
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func createWebhookService(context *clusterd.Context) error {
+	webhookService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      appName,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Port: servicePort,
+					TargetPort: intstr.IntOrString{
+						IntVal: serverPort,
+					},
+				},
+			},
+
+			Selector: map[string]string{
+				"app": "rook-ceph-admission-controller",
+			},
+		},
+	}
+	_, err := context.Clientset.CoreV1().Services(namespace).Create(&webhookService)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+// InitWebhookIfSecretPresent will initialize the webhook if secret is detected
+func InitWebhookIfSecretPresent(context *clusterd.Context, admissionImage string) error {
+	isPresent, err := isSecretPresent(context)
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve secret")
+	}
+	if isPresent {
+		err = initWebhook(context, admissionImage)
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize webhook")
+		}
+	}
+	return nil
+}
+
+func initWebhook(context *clusterd.Context, admissionImage string) error {
+	// At this point volume should be mounted, so proceed with creating the service and validatingwebhookconfig
+	err := createWebhookService(context)
+	if err != nil {
+		return errors.Wrap(err, "failed to create service")
+	}
+	err = createWebhookPods(context, admissionImage)
+	if err != nil {
+		return errors.Wrap(err, "failed to create service")
+	}
+	return nil
+}
+
+func createWebhookPods(context *clusterd.Context, admissionImage string) error {
+	admission_parameters := []string{"admission-controller",
+		"--storage", storageFlag}
+	secretVolume := getSecretVolume()
+	secretVolumeMount := getSecretVolumeMount()
+
+	antiAffinity := csi.GetPodAntiAffinity("app", "rook-ceph-admission-controller")
+	admissionControllerDeployment := getDeployment(secretVolume, antiAffinity, admissionImage, admission_parameters, secretVolumeMount)
+
+	_, err := context.Clientset.AppsV1().Deployments(namespace).Create(&admissionControllerDeployment)
```

Comment:
Here you will need to update the deployment if it already exists in case the spec changed. 

Because of this, right now if you restart the operator it is failing with this error:
```
: failed to start webhook: failed to initialize webhook: failed to create service: failed to create admission controller replica set: deployments.apps "rook-admission-controller" already exists
```

Reply:
Calling `k8sutil.CreateDeployment()` should take care of the update for you.

Reply:
didnt know this. updated ! Thanks